src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -1095,6 +1095,17 @@ void G1CollectedHeap::do_full_collection(bool clear_all_soft_refs) {
                                   do_maximum_compaction);
 }
 
+bool G1CollectedHeap::upgrade_to_full_collection() {
+  log_info(gc, ergo)("Attempting full compaction clearing soft references");
+  bool success = do_full_collection(false /* explicit gc */,
+                                    true  /* clear_all_soft_refs */,
+                                    false /* do_maximum_compaction */);
```

Comment:
Shouldn't this have `do_maximum_compaction` == true?  That's what the change to `VM_G1CollectForAllocation` expects.  In which case, do we really need separate arguments for `do_full_collection` for `clear_all_soft_refs` and `do_maximum_compaction` or could they be collapsed to one `maximum_compaction` argument?

Reply:
This is currently the only case where we need separate arguments for the two. So previously the "upgraded" Full collections used `clear_all_soft_refs == true` and `do_maximum_compaction == false`, so I kept it that way. In some sense I think this is fine, when upgrading this is the first time we try a Full GC so leaving dead wood (`maximum==false`) could still find some space. It is only when we attempt a second Full collection in a row we set `do_maximum_compaction == true`. Makes sense, or do you think we should force a "maximum compaction" when upgrading?

Reply:
Fwiw, both is fine with me. The literal translation would certainly be to do a maximally compacting collection here, but as @kstefanj mentioned, the non-maximally compacting collection may be faster and in most cases sufficient.
I would suggest to file an RFE (which I'll do later) to try to consolidate all these paths in a separate change, and in this change just remove the third "upgraded" gc.

Reply:
Fwiw, I am good with either, i.e. do a maximally compacting gc or not; both will likely free some space but the non-maximally compacting should be faster. Maybe this change should just remove the third full gc, and in a separate change think about the upgrade policy. I already filed https://bugs.openjdk.java.net/browse/JDK-8268393 for such work.

Reply:
At least with this change we will never to 3 consecutive Full GCs in the same VM operation.